### ISSUE TITLE
Ctrl+C only quits a program after any other key is pressed

### ISSUE THREAD

**Issue by doonv** [pyreadline3/pyreadline3#10]
## <kbd>Ctrl+C</kbd> only quits a program after any other key is pressed. 
This always takes effect when `readline` is imported. Here is a little example program.
```py
import readline
readline.parse_and_bind("tab: complete")

def complete(text,state):
    volcab = ['dog','cat','rabbit','bird','slug','snail']
    results = [x for x in volcab if x.startswith(text)] + [None]
    return results[state]


readline.set_completer(complete)

try:
    line = input('prompt> ')
except KeyboardInterrupt:
    from os import system
    system("color")
    print("\n\033[91mKeyboardInterrupt\033[0m")
```
Press <kbd>Ctrl+C</kbd>, you will see that nothing happens until you press a regular key like <kbd>H</kbd>.  This is on Bindows 10.

**Comment by davedikian** [comment]
I noticed this issue as well and managed to trace it to the "getkeypress" function in console.py. Adding an additional conditional to break out of the polling loop after receiving a `Ctrl+C` like there currently is for `Ctrl+S` seems to be sufficient to immediately propagate the exception.

```python
def getkeypress(self):
    '''Return next key press event from the queue, ignoring others.'''
    while True:
        e = self.get()
        if e.type == 'KeyPress' and e.keycode not in key_modifiers:
            log("console.getkeypress %s" % e)
            if e.keyinfo.keyname == 'next':
                self.scroll_window(12)
            elif e.keyinfo.keyname == 'prior':
            self.scroll_window(-12)
            else:
                return e
        elif ((e.type == 'KeyRelease') and
              (e.keyinfo == KeyPress('S', False, True, False, 'S') or
               e.keyinfo == KeyPress('C', False, True, False, 'C'))): # additional condition for Ctrl+C
            log("getKeypress:%s,%s,%s" % (e.keyinfo, e.keycode, e.type))
            return e
```

I haven't really tested it that much, but it seems to work properly now. I feel like this shouldn't be necessary to immediately propagate the exception, but I don't know enough about the implementation to question it further.

**Comment by brgirgis** [comment]
@davedikian, please create a PR

**Comment by davedikian** [comment]
@brgirgis Done, created a new PR for this change: https://github.com/pyreadline3/pyreadline3/pull/26

**Comment by brgirgis** [comment]
released in 3.5.0

**Comment by Zeta314** [comment]
Hi everyone,
I'd like to report that the issue is resolved only when running the script via cmd.exe. However, the problem persists when using powershell.exe.

**Comment by davedikian** [comment]
@Zeta314 I primarily use Powershell for development, meaning this was the first thing I checked when debugging this issue and then testing the fix, so I'm not sure what could be causing the issue to persist. What Python and Powershell versions are you using? Are you using the above test script to verify that this is still occurring, or something else?

**Comment by Zeta314** [comment]
I use:
- python 3.12.5
- windows terminal (as host when not in vscode)
- poetry 1.8.3
- powershell 7.4.5

I am using this snippet of code:
```python
should_stop = False
            while not should_stop:
                try:
                    line = input(self.prompt).strip()
                except KeyboardInterrupt:
                    print("", file=self.stdout)
                    continue
                except EOFError:
                    should_stop = True
                    continue

                if not line:
                    continue

                command, args = self.__split_line(line)
                should_stop = not self.__handle_command(command, args)
```

**Comment by davedikian** [comment]
@Zeta314 I'm also using Powershell 7.4.5 and Python 3.12, so I don't think it's a version issue then. If it's still requiring an additional key press, it's likely still localized to the same area I changed before. As long as you're on the latest version of pyreadline3, the only debug method I can think of is for you to collect some logs and share them.

If that sounds good, you can enable debug logging by removing or commenting out the line `log = nolog` in any file of interest (in this case I'd like to see the logs from `console/console.py`). Then, add the following lines to the top of any script you're using:

```python
import pyreadline3
pyreadline3.logger.start_file_log("debug.log") # Or whatever filename you prefer
```

Once you've replicated this issue, I'd be interested to see the contents of the resulting log. The default logging statement placement does not log every single key event however, so if you want to go the extra mile, duplicating line 561 to underneath 548 will also log everything you press, even if it does not trigger the `getkeypress` logic. It should look like this:

```python
while True:
    e = self.get()
    log("getKeypress:%s,%s,%s" % (e.keyinfo, e.keycode, e.type)) # probably should adjust the message to make it clear it's logging everything
```
etc.

Outside of that, I'm at a loss as to why this could still be an issue.

**Comment by Zeta314** [comment]
I've done everything you said, here's the log file:

[debug.log](https://github.com/user-attachments/files/17029967/debug.log)

This is the code i tested it with (basically the same as my code):
```python
import readline
import pyreadline3

pyreadline3.logger.start_file_log("debug.log")


def complete(self, text: str, state: int) -> str | None: ...


old_completer = readline.get_completer()
readline.set_completer(complete)
readline.parse_and_bind("tab: complete")

try:
    should_stop = False
    while not should_stop:
        try:
            line = input("(cmd) ").strip()
        except KeyboardInterrupt:
            print()
            continue
        except EOFError:
            should_stop = True
            continue

        if not line:
            continue

        print(f"line: {line}")
finally:
    readline.set_completer(old_completer)
```

**Comment by Zeta314** [comment]
@davedikian I've tried starting it with `powershell.exe -NoProfile`, however the issue still resides. I'm not entirely sure what's causing it though.

**Comment by davedikian** [comment]
@Zeta314 I've developed an alternate solution that checks for a waiting event before getting blocked by the `ReadConsoleInputW` call. For my testing, I disabled my original solution and this ended up also fixing the issue, now without relying on a keyboard event. See if implementing the following changes into `console.py` fixes your issue as well:

1.) Add the following to the `funcs` list found on line 159:
```python
"GetNumberOfConsoleInputEvents",
```

2.)  Replace the method `get` around line 532 with the following contents:
``` python
def get(self):
    """Get next event from queue."""
    inputHookFunc = c_void_p.from_address(self.inputHookPtr).value

    Cevent = INPUT_RECORD()
    count = DWORD(0)
    while True:
        event_count = DWORD(0)
        if inputHookFunc:
            call_function(inputHookFunc, ())
        self.GetNumberOfConsoleInputEvents(self.hin, byref(event_count))
        if event_count:
            status = self.ReadConsoleInputW(self.hin, byref(Cevent), 1, byref(count))
            if status and count.value == 1:
                e = event(self, Cevent)
                return e
```

3.) Finally, add the following defintions underneath around line 696:
```python
Console.GetNumberOfConsoleInputEvents.restype = BOOL
Console.GetNumberOfConsoleInputEvents.argtypes = [HANDLE, LPDWORD]
```

(Line numbers are based on where things end up after all the additions.)

Let me know if this fixes your issue or if anything else happens. If it works, I'll do some more testing and see about putting it into another PR.

**Comment by Zeta314** [comment]
@davedikian perfect! With the patch you've provided it all seems to work. Thank you!
Looking forward for the PR & new version!
